File: config/config-annotation-processor/src/test/java/ru/tinkoff/kora/config/annotation/processor/AbstractConfigTest.java
package ru.tinkoff.kora.config.annotation.processor;

import org.intellij.lang.annotations.Language;
import ru.tinkoff.kora.annotation.processor.common.AbstractAnnotationProcessorTest;
import ru.tinkoff.kora.config.annotation.processor.processor.ConfigParserAnnotationProcessor;
import ru.tinkoff.kora.config.annotation.processor.processor.ConfigSourceAnnotationProcessor;
import ru.tinkoff.kora.config.common.extractor.ConfigValueExtractor;

import java.lang.reflect.InvocationTargetException;
import java.util.List;

public abstract class AbstractConfigTest extends AbstractAnnotationProcessorTest {

    @Override
    protected String commonImports() {
        return super.commonImports() + """
            import ru.tinkoff.kora.config.common.annotation.*;
            """;
    }

    protected ConfigValueExtractor<Object> compileConfig(List<?> arguments, @Language("java") String... sources) {
        super.compile(List.of(new ConfigParserAnnotationProcessor(), new ConfigSourceAnnotationProcessor()), sources);
        this.compileResult.assertSuccess();
        var args = arguments
            .stream()
            .map(o -> o instanceof GeneratedResultCallback<?> c ? c.get() : o)
            .toArray();

        try {
            return (ConfigValueExtractor<Object>) this.compileResult.loadClass("$TestConfig_ConfigValueExtractor")
                .getConstructors()[0]
                .newInstance(args);
        } catch (InstantiationException | IllegalAccessException | InvocationTargetException e) {
            throw new RuntimeException(e);
        }
    }
}

File: config/config-symbol-processor/src/test/kotlin/ru/tinkoff/kora/config/symbol/processor/AbstractConfigTest.kt
package ru.tinkoff.kora.config.symbol.processor

import org.intellij.lang.annotations.Language
import ru.tinkoff.kora.config.common.extractor.ConfigValueExtractor
import ru.tinkoff.kora.config.ksp.processor.ConfigParserSymbolProcessorProvider
import ru.tinkoff.kora.config.ksp.processor.ConfigSourceSymbolProcessorProvider
import ru.tinkoff.kora.ksp.common.AbstractSymbolProcessorTest

abstract class AbstractConfigTest : AbstractSymbolProcessorTest() {
    override fun commonImports(): String {
        return super.commonImports() + """
            import ru.tinkoff.kora.config.common.annotation.*;
        """.trimIndent()
    }

    protected open fun compileConfig(arguments: List<*>, @Language("kotlin") vararg sources: String): ConfigValueExtractor<Any?> {
        super.compile(listOf(ConfigSourceSymbolProcessorProvider(), ConfigParserSymbolProcessorProvider()), *sources)
        compileResult.assertSuccess()
        return compileResult.loadClass("\$TestConfig_ConfigValueExtractor")
            .constructors[0]
            .newInstance(*arguments.map { if (it is GeneratedObject<*>) it() else it }.toTypedArray()) as ConfigValueExtractor<Any?>
    }
}
